geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LPopDUnitTest.java

```diff
@@ -0,0 +1,139 @@
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.redis.internal.commands.executor.list;
+
+import static org.apache.geode.test.dunit.rules.RedisClusterStartupRule.BIND_ADDRESS;
+import static org.apache.geode.test.dunit.rules.RedisClusterStartupRule.REDIS_CLIENT_TIMEOUT;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+
+import org.apache.geode.redis.ConcurrentLoopingThreads;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+
+public class LPopDUnitTest {
+  private static final String KEY_BASE = "key";
+  private static final int NUM_LISTS = 100;
+  private static final String ELEMENT_BASE = "element-";
+  public static final int NUM_VMS = 4;
+
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule(4);
+
+  private static final int LIST_SIZE = 100;
+  private static JedisCluster jedis;
+
+  @BeforeClass
+  public static void classSetup() {
+    MemberVM locator = clusterStartUp.startLocatorVM(0);
+    clusterStartUp.startRedisVM(1, locator.getPort());
+    clusterStartUp.startRedisVM(2, locator.getPort());
+    clusterStartUp.startRedisVM(3, locator.getPort());
+
+    int redisServerPort = clusterStartUp.getRedisPort(1);
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort), REDIS_CLIENT_TIMEOUT);
+  }
+
+  @Before
+  public void testSetup() {
+    clusterStartUp.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis.close();
+  }
+
+  @Test
+  public void shouldDistributeDataAmongCluster_andRetainDataAfterServerCrash()
+      throws Exception {
+    List<String> elementList1 = makeElementList(0, LIST_SIZE / 2, ELEMENT_BASE);
+    List<String> elementList2 = makeElementList(LIST_SIZE / 2, LIST_SIZE, ELEMENT_BASE);
+    List<String> allElements = makeElementList(0, LIST_SIZE, ELEMENT_BASE);
+
+    new ConcurrentLoopingThreads(NUM_LISTS / 2,
+        (i) -> jedis.lpush(KEY_BASE + i, elementList1.toArray(new String[] {})),
+        (i) -> jedis.lpush(KEY_BASE + i, elementList2.toArray(new String[] {})),
+        (i) -> jedis.lpush(KEY_BASE + (i + NUM_LISTS / 2), elementList1.toArray(new String[] {})),
+        (i) -> jedis.lpush(KEY_BASE + (i + NUM_LISTS / 2), elementList2.toArray(new String[] {})))
+            .runInLockstep();
+
+    clusterStartUp.crashVM(NUM_VMS - 1);
+
+    confirmAllDataIsPresent(allElements);
+  }
```

Comment:
This test is not testing the distributed behaviour of LPOP, but rather testing that secondary copies of buckets contain the same data as primary copies after concurrent LPUSH calls, so that when we failover due to the server crash, data is consistent. As such, this test should be reworked to test distributed LPOP behaviour. Examples of what we could test:
- With two servers, loading a list on one server, calling LPOP repeatedly to empty half the list, then crashing the primary server results in the list containing what we expect it to after the secondary is promoted to primary
- With at least three servers, loading a list, then concurrently calling LPOP while forcing the primary to move to a member that doesn't host the bucket (similar to what's done in `StringsDUnitTest.givenBucketsMoveDuringAppend_thenDataIsNotLost`) and asserting that we never get the same element returned from LPOP twice (which would indicate that we did an LPOP without actually removing the element).

Given that LPOP can suffer from the same problems as APPEND, where a retried command results in the operation being performed twice, care needs to be taken with what we assert in the second test case so as not to produce a flaky test.

Reply:
Have a couple tests along those lines now, instead. Had to use three servers for both; in the first one, if you crash the second, it fails because it can't ensure required redundancy.